Layering for atmospheric pipes

chromiumboy opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
This PR adds layering for atmospheric pipes - up to three pipes of the same type/orientation can now be built on a single tile (see the demonstration video below)

All pipes and atmospheric devices, when constructed, will be assigned to the default layer (referred to as the 'primary configuration'). Prototypes for the two alternative configurations for pipes and most atmospheric devices have been added for the benefit of mappers

To change the configuration of a pipe or atmospheric device, right-click on the entity while holding a screwdriver, open the 'Adjust' submenu, and select the new configuration. Pipes and atmospheric devices can be adjusted while anchored, unanchored, or being held. If the entity is being held, configurations can be cycled through using the 'z' key. The majority of atmospheric devices can be assigned to any of the three configurations with the exception of radiators (I wasn't satisfied with the sprites I tried to make for the alternative configurations...). Note that if a player tries to switch a pipe to a sublayer that is occupied, the pipe will become unanchored

This PR also adds the pipe manifold, a constructible multi-pronged pipe that connects pipes across sublayers

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
This is a requested feature that was ported from SS13. It will allow for additional flexibility/complexity in station atmospheric setups

## Technical details
<!-- Summary of code changes for easier review. -->
The PR is relatively straightforward. The logic for the pipe layering is handled by the AtmosPipeLayersSystems and the associated component. The bulk of the changes are due to the addition of new sprites and yaml alterations
 
## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
[Demonstration of pipe layering in action](https://youtu.be/b10nKdeLviY)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->
N/A

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- add: Atmospheric pipes can now be built on three different sublayers, allowing for multiple pipe networks on a single tile
- add: Pipe manifolds can now be constructed to connect pipes across different sublayers

keronshb commented:
Finally, atmos mains eating GOOD.

Partmedia commented on Content.Shared/Atmos/Piping/PipeColorVisuals.cs:
This doesn't look like it needed to be changed here and just increases the diff size and risk of conflicts.

Partmedia commented on Resources/Locale/en-US/verbs/verb-system.ftl:
I feel like we can come up with a better name besides "Adjust", but it's also fine as-is.

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/alt_layers.yml:
```diff
@@ -0,0 +1,717 @@
+      map: [ "enum.PipeVisualLayers.Pipe" ]
+      renderingStrategy: Default
+
+# GasThermoMachineHeater
```
Why do thermomachines need variants for all different layers? Can't you get a layered T to adapt from e.g. a third position to a primary position, and branch that off to this thermomachine?

I'm concerned that:

1. This bloats too many entitiy prototypes
2. This contributes to "messy layers" which is one of the reasons why others have been opposed to pipe layers

Partmedia replied:
Ditto for a lot of the machines that follow.

chromiumboy replied:
I tried making an automatic placement system, but the solution was probably over-engineered and [the result was a bit jittery](https://youtu.be/sIdewKx9Q3w)

Going to stick with the separate entities prototypes for each layer now, for the pipes at least - I removed the alternative prototypes for the atmos devices

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/trinary.yml:
This is a refactoring change that is blended in with this large PR.

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/unary.yml:
```diff
@@ -189,14 +189,14 @@
           map: [ "enum.LightLayers.Unshaded" ]
           color: "#990000"
+    - type: Appearance 
     - type: GenericVisualizer
       visuals:
        # toggle color of the unshaded light:
        enum.OutletInjectorVisuals.Enabled:
          enum.LightLayers.Unshaded:
            True: { color: "#5eff5e" }
            False: { color: "#990000" }
-    - type: Appearance
```
Changes like this really make reviewing annoying...

Partmedia reviewed (commented):
I did an initial first-pass review of primarily the changes to `PipeNode` and the addition of `AtmosPipeLayersComponent`. I did not review anything to do with the atmos monitoring console but since you wrote that, I trust what you did there is fine. Will come back to check `PipeRestrictOverlapSystem` and `GasPipeManifoldSystem`. I did not look at any sprites. Will give it a more extensive in-game test soon.

Overall:

- Thanks for working on this.
- I think the changes to core data structures are sound and well designed.
- Could you investigate whether it would make sense to have `AtmosPipeLayersComponent` rolled into another class? For example, if you wanted to make a special kind of pipe (e.g. radiator) that only supports one layer, that would be the "trivial" case of filling out everything in `AtmosPipeLayersComponent` except only with one layer, right? Or, is this already how this is used?
- I'm concerned about the number of machines with multi-layer variants. As I mentioned in the in-line comments, I think this contributes too much to prototype bloat and also might encourage "messy" atmos designs that are hard to understand. I would personally prefer to have most machines in primary configurations that you would T out. I think some past maintainers that have chimed in about this matter have said something similar, though I would have to dig out the context to confirm.

metalgearsloth commented:
With disposals PR can probably predict most of this so will work on getting the reviews done for that soon.

ArtisticRoomba commented:
my mouth is watering and my brain is filled with the bottomless possibilities of 3 layers to play with. unreal.

Minemoder5000 commented:
Would it be possible to add cycling through pipe layers in the construction menu? Maybe hijack the keybind for flipping, or something similar?

LaCumbiaDelCoronavirus commented:
Do pipe layers have to be switched between layers when pipes are placed, or are there multiple variants for each layer in the construction menu?

ScarKy0 commented:
And it's not even roomba's birthday

VerinSenpai commented:
oh my god. its... its.... *dies*

Partmedia commented:
While I understand the excitement about this, please keep discussions on GitHub on-topic and review-related. This isn't a chat room. Thanks.

VerinSenpai commented:
if it hasn't been done already, would it be possible to include a hot key for switching between the states, similar to how rotating is done?

TurboTrackerss14 commented:
Gunna let the real atmos code gamers manage the code side. 
But the readability of some of these pipes is... not great. 

Canister covering port can/will lead to incorrect layering. 
![image](https://github.com/user-attachments/assets/2abc99c7-25f0-425d-82ba-d20308f41c56)

Pipelayers resetting to "center" to connect to devices will lead to "why are these two pipes not connected, they line up" ahelps
![image](https://github.com/user-attachments/assets/b7ecd290-64ab-4305-a0a8-759fb829d28f)

Overlapping manual valves handles will also lead to "its not active why it pump" ahelps
![image](https://github.com/user-attachments/assets/4111834a-7da7-43e8-9d21-ec6b92a84d4e)

Auto linking pipes, bad awful, something will explode surely. (maybe only allow layering changes on unanchored?)
![image](https://github.com/user-attachments/assets/8accdcb8-a06f-4c4d-a222-9c8e41756a95)

chromiumboy commented:
> Would it be possible to add cycling through pipe layers in the construction menu? Maybe hijack the keybind for flipping, or something similar?

I considered this, but you'd need a separate construction graph for each layer for every pipe or atmos device, and I didn't want to do that :P It would also bloat the construction menu, or require that the menu be updated to support layers. Best bet is to implement the RPD, it should be used for large scale pipe net constructions in any case

chromiumboy commented:
Sounds like a bug, I'll have a look. All configs should need a short do after to cycle them

keronshb commented:
Wouldn't some of the changes like changing the gas pipe sensor parent from GasPipeBase > GasBinaryBase and changing the pipe states from pipeStraight > pipeBinaryConnector count as breaking changes? Otherwise looks fine to me.

chromiumboy commented:
> Wouldn't some of the changes like changing the gas pipe sensor parent from GasPipeBase > GasBinaryBase and changing the pipe states from pipeStraight > pipeBinaryConnector count as breaking changes? Otherwise looks fine to me.

I'm not sure how these changes might potentially break something. Did you have something in mind?

ArtisticRoomba reviewed (approved):
The YAML and RSI side looks good to me.

I've left some comments on the C# side of the PR (remember that I don't know much), so feel free to take them into consideration or disregard if necessary.

Partmedia commented on Content.Shared/Atmos/Components/AtmosPipeLayersComponent.cs:
```diff
@@ -0,0 +1,98 @@
+    /// (from 0 to <see cref="NumberOfPipeLayers"/> - 1).
+    /// </remarks>
+    [DataField]
+    public string[]? SpriteRsiPaths = null;
```
Why make it nullable at all? Why not a `string[]` that is left empty if not specified?

chromiumboy commented:
> The YAML and RSI side looks good to me.
> 
> I've left some comments on the C# side of the PR (remember that I don't know much), so feel free to take them into consideration or disregard if necessary.

Thanks for the review :)

Partmedia commented:
Perhaps you missed this in the last pass, but why does `public string[]? SpriteRsiPaths;` have to be nullable Wouldn't checking that it is a zero-length list be the same?

chromiumboy commented:
> Perhaps you missed this in the last pass, but why does `public string[]? SpriteRsiPaths;` have to be nullable? Wouldn't checking that it is a zero-length list be the same?

The idea was that a length zero array would indicate that it hadn't been initiated correctly, while a null array meant it was never initialized at all. But I guess it doesn't really matter in this context? I can change it

Partmedia commented:
> > Perhaps you missed this in the last pass, but why does `public string[]? SpriteRsiPaths;` have to be nullable? Wouldn't checking that it is a zero-length list be the same?
> 
> The idea was that a length zero array would indicate that it hadn't been initiated correctly, while a null array meant it was never initialized at all. But I guess it doesn't really matter in this context? I can change it

I'm not sure what the distinction between these two states are: both of them result in the same condition of "layer sprites should not be used" in the code, right? There's no extra handling between the two cases?

chromiumboy commented:
> > > Perhaps you missed this in the last pass, but why does `public string[]? SpriteRsiPaths;` have to be nullable? Wouldn't checking that it is a zero-length list be the same?
> > 
> > 
> > The idea was that a length zero array would indicate that it hadn't been initiated correctly, while a null array meant it was never initialized at all. But I guess it doesn't really matter in this context? I can change it
> 
> I'm not sure what the distinction between these two states are: both of them result in the same condition of "layer sprites should not be used" in the code, right? There's no extra handling between the two cases?

Yeah, there's no difference in how they are handled here

southbridge-fur commented:
It's beautiful

Partmedia commented:
Instead of normal, alt 1, and alt 2, I think a better set of names would be Layer 1, Layer 2, and Layer 3:
![image](https://github.com/user-attachments/assets/84be794b-6049-44f3-beef-f669e948d94a)

Then you can make it match here:
![image](https://github.com/user-attachments/assets/f87c3208-1317-4c2c-a5f9-b3328ace24f1)

It would be nice if the aghost could skip the screwdriver check to adjust. Perhaps "Adjust" should be renamed to "Adjust Layer".

Some sprites still need adjustment to correctly show where the pipes connect:
![image](https://github.com/user-attachments/assets/ba158578-c4f2-48e0-8625-92b5e7c179d6)

But I understand that this is a lot of work and not necessarily in scope for this PR.

I think the interface for constructing the pipes via the construction menu was a bit clunky and liked the interface you showed earlier with the layer snapping.
![image](https://github.com/user-attachments/assets/58353fb3-0d64-4d7a-9a0e-9f9449e5d70d)

Partmedia commented:
Changing the `CurrentPipeLayer` in the `AtmosPipeLayerComponent` does not cause the visualizer to correctly update.

chromiumboy commented:
> Instead of normal, alt 1, and alt 2, I think a better set of names would be Layer 1, Layer 2, and Layer 3
> 
> Then you can make it match here:

Refering to them as different 'layers' might be ambiguous for new players, and using a linear numbering system seems a bit weird since the first layer starts with the pipe in the middle of the tile. Happy to debate the merits of either approach though

> It would be nice if the aghost could skip the screwdriver check to adjust. Perhaps "Adjust" should be renamed to "Change Layer".

Agreed that aghost should skip the tool check. I thought that having a generic 'adjust' verb might be useful if someone wanted to reuse it, as 'change layers' is very specific

> Some sprites still need adjustment to correctly show where the pipes connect
>
> But I understand that this is a lot of work and not necessarily in scope for this PR.

It should be simple, the yaml prototype likely just needs updating

> I think the interface for constructing the pipes via the construction menu was a bit clunky and liked the interface you showed earlier with the layer snapping.

I can add it back in if you wanna test it out and see what you prefer

chromiumboy commented:
> Changing the `CurrentPipeLayer` in the `AtmosPipeLayerComponent` does not cause the visualizer to correctly update.
> 
> If the only intended way for this to be adjusted is the right click menu, then the VV field should be set to read-only.

Yeah, it's not intended to be changed via VV

Partmedia commented:
> Refering to them as different 'layers' might be ambiguous for new players, and using a linear numbering system seems a bit weird since the first layer starts with the pipe in the middle of the tile. Happy to debate the merits of either approach though

Going middle, left, right for "Primary", "Secondary", and "Tertiary" seem very confusing and easy to forget to me.

Especially since there's a mix of two different types of names, e.g. "Secondary" but "1st Alt" in the entity spawn menu.

A linear ordering still makes more sense to me, but I suppose you can wait for some other input and see if we can come to a consensus on what the best organization might be.

ArtisticRoomba commented:
Did an in-game test, overall, it went well!
![Content Client_9kyFHxyPpU](https://github.com/user-attachments/assets/041b87cc-ed16-4bdb-b141-59b6dfe88699)
![image](https://github.com/user-attachments/assets/8b42714f-a4e4-4af5-8755-22363061ad17)

As you can see, there appears to be some visual bugs with the volumetric pump.

Additionally, using a gas analyzer on any manifold is also quite a pain:

![image](https://github.com/user-attachments/assets/3eee3296-2bc5-4a44-bf4e-7d2bec457717)

It seems to keep adding another pipenode view onto the next one every time it updates.

Feedback from things that I observed, in no particular order:
- Building pumps in different configs is quite annoying as you have to switch their configuration and then anchor them in order to get them where you want.
- You can't stack 3 pumps on one tile, each in their different configuration.
- Looking at the verb to switch the configuration every single time is really, really slow. Maybe an alt-click interaction with a screwdriver would be better?
	- Doing this with a drill is even slower as you sometimes forget to have the drill in screwing mode, so when you go through the work to look at the "adjust" verb it says you can't do anything and you have to look at the verb all over again.
- Pipes constantly give the construction hint saying they can be turned into their different variants (seems to be only the straight variant).

Overall, what an amazing system! It really works well, I didn't really have any major problems with tinkering around with making a new TEG design.

chromiumboy commented:
Cooking some more on this

ArtisticRoomba commented:
Tried out the contextual pipe placement for a bit, was extremely cool!

![Content Client_EeQLVtzfyN](https://github.com/user-attachments/assets/363ba6cd-b5a8-43a8-a82e-587b27f36d26)

Seemed a bit weird behaving originally but the technique is to move your mouse diagonally. A possible suggestion is to draw a diagonal yellow line that indicates layer placements?

A bug I observed is that every time I made a layer change it seemed to cancel whatever keyboard input I was doing. So walking around became stutterey when I moved my mouse to move the pipe ghost between different pipe layers.

chromiumboy commented:
> A bug I observed is that every time I made a layer change it seemed to cancel whatever keyboard input I was doing. So walking around became stutterey when I moved my mouse to move the pipe ghost between different pipe layers.

It looks like it's an engine side issue, I'll need to make a separate PR to address that. I should also make a PR that will allow multiple pipe construction ghosts to be present on the same tile - at the moment, the system is limited to one per tile

I made some changes to how manifolds look - they now only show their connector as needed (including on the atmos monitor)

![manifold](https://github.com/user-attachments/assets/6d4be6c5-dee4-4a4c-994a-f5822158067b)   ![manifold-map](https://github.com/user-attachments/assets/a3ac86a8-e726-4e45-896f-156611a57f5a)

VerinSenpai commented:
now THAT is beautiful.

ArtisticRoomba commented:
Did another playtest, was extremely fun!
![image](https://github.com/user-attachments/assets/8e698f7f-333d-4434-8ea8-f003c799ebaf)

Some feedback, in no particular order:
- The contextual pipe placement system is REALLY powerful. Like, I built the above image very quickly with very little hiccups at all.
- The auto-adjust-config on screwdriver use is really powerful and allows you to switch up your setups really fast. It's no longer clumsy to use.
- Attempting to adjust a vent that is in pressure lockout will start two do-afters, one for adjusting the configuration of the device, and the other for releasing the lockout.
- The ability to adjust configurations on devices that are under tiles (ex. air vents, scrubbers) is possible, which is a problem.
- Building large straight lines of pipe using CTRL is no longer possible.
- Building pumps in different configurations is still a bit finicky, I have to adjust the config before placing it down, which makes it a hassle for tight spaces. Ideally all devices that have pipe layering options can be placed down in the configuration you want in the ghost builder using the 3-dots guide (scope inflator 9000 though).

Oh I love it.

Partmedia commented on Content.Shared/Atmos/Components/AtmosPipeLayersComponent.cs:
```diff
@@ -0,0 +1,110 @@
+    /// If null, the RSI path does not differ between pipe layers.
+    /// </summary>
+    /// <remarks>
+    /// If not null, it is assumed that there will be an array entry for each pipe layer
+    /// (from 0 to <see cref="NumberOfPipeLayers"/> - 1).
```
This comment is no longer relevant.

Qiler commented:
If you move a pipe to a layer that is occupied, it will unwrench the that pipe but wont remove gasses from it.

chromiumboy commented:
> If you move a pipe to a layer that is occupied, it will unwrench the that pipe but wont remove gasses from it.

Good catch, I'll fix that. Thanks for letting me know

Partmedia commented:
In-game test looks good to me. Will hopefully get back to make another pass at the code soon.

Dicerson commented:
Can we see an example of what the layering looks like for actual devices, and not just pipes? For example stacked filters, radiators, ports, scrubbers, etc.? Especially with adjacent pipes or manifolds to see how all those connections lay out?

ArtisticRoomba commented:
> Can we see an example of what the layering looks like for actual devices, and not just pipes? For example stacked filters, radiators, ports, scrubbers, etc.? Especially with adjacent pipes or manifolds to see how all those connections lay out?

I have provided plenty of examples if you look at the PR's comments.

chromiumboy commented:
> > Can we see an example of what the layering looks like for actual devices, and not just pipes? For example stacked filters, radiators, ports, scrubbers, etc.? Especially with adjacent pipes or manifolds to see how all those connections lay out?
> 
> I have provided plenty of examples if you look at the PR's comments.

I stole one of your examples and added it to the media section of the PR description ;)

BTW, do you have any time coming up soon to do that demonstration video?

ArtisticRoomba commented:
Oh, sure, I'll get that done! I was crammed with finals for the previous 2 weeks and Thursday is my last final, so I'll be more free soon. I spent today and yesterday reviewing Xenoborgs and mapping PRs. So I'll have a little video out for you in the coming days.

Partmedia commented:
> Just following up to see how it's going. It would be nice to have this PR ready to merge by the weekend so that there will be time to test it on Vulture before the staging cut-off next week

Sorry about the delay; I'm currently traveling and so it is unlikely I'll be able to finish looking by the weekend. If anyone else is able to complete a code review, that would be great.

Dicerson commented:
> > Can we see an example of what the layering looks like for actual devices, and not just pipes? For example stacked filters, radiators, ports, scrubbers, etc.? Especially with adjacent pipes or manifolds to see how all those connections lay out?
> 
> I have provided plenty of examples if you look at the PR's comments.

I see only a single image providing examples, and in that image there are no instances of filters stacked on the same tile from different layers, nor any kind of stacking between devices that aren't radiators (which can already be stacked on top of stuff currently). The video also does not show the stacking of anything that isn't just pipes.

Am I to take it that only pure pipes benefit from this, and that filters and other devices are not included in the layering?

chromiumboy commented:
> > > Can we see an example of what the layering looks like for actual devices, and not just pipes? For example stacked filters, radiators, ports, scrubbers, etc.? Especially with adjacent pipes or manifolds to see how all those connections lay out?
> > 
> > 
> > I have provided plenty of examples if you look at the PR's comments.
> 
> I see only a single image providing examples, and in that image there are no instances of filters stacked on the same tile from different layers, nor any kind of stacking between devices that aren't radiators (which can already be stacked on top of stuff currently). The video also does not show the stacking of anything that isn't just pipes.
> 
> Am I to take it that only pure pipes benefit from this, and that filters and other devices are not included in the layering?

Correct, you can still only have one device per tile

Partmedia commented on Content.Server/Atmos/EntitySystems/PipeRestrictOverlapSystem.cs:
```diff
@@ -97,27 +97,27 @@ public bool CheckOverlap(Entity<NodeContainerComponent, TransformComponent> ent)
+            foreach (var (otherDir, otherLayer) in otherDirsAndLayers)
             {
-                if ((dir & otherDir) != 0)
+                if ((dir & otherDir) != 0 && layer == otherLayer)
```
is `!= 0` needed?

chromiumboy replied:
It's required, it checks whether the two pipes have nodes with the same orientation. My modification adds that the pipe nodes also have to be on the same layer to count as an overlap

Partmedia commented on Content.Server/Atmos/Piping/EntitySystems/AtmosPipeAppearanceSystem.cs:
```diff
@@ -34,9 +34,12 @@ private void UpdateAppearance(EntityUid uid, AppearanceComponent? appearance = n
         if (!TryComp<MapGridComponent>(xform.GridUid, out var grid))
             return;
 
+        var numberOfPipeLayers = TryComp<AtmosPipeLayersComponent>(uid, out var atmosPipeLayers) ? atmosPipeLayers.NumberOfPipeLayers : 1;
```
This got repeated again here.

Partmedia commented on Content.Server/Atmos/Piping/EntitySystems/GasPipeManifoldSystem.cs:
Why does a manifold behave any differently compared to a 6-way pipe?

chromiumboy replied:
The manifold connects across different pipe layers; all other basic pipes do not

Partmedia commented on Content.Server/NodeContainer/Nodes/PipeNode.cs:
```diff
@@ -24,6 +24,12 @@ public partial class PipeNode : Node, IGasMixtureHolder, IRotatableNode
+        ///     The *current* layer to which the pipe node is assigned.
+        /// </summary>
+        [DataField("pipeLayer")]
+        public byte CurrentPipeLayer;
```
I wonder if this should remain a `byte`, or if it should gain a new `PipeLayer` type (like the above `PipeDirection`).

Partmedia commented on Content.Shared/Atmos/Components/AtmosPipeLayersComponent.cs:
```diff
@@ -0,0 +1,110 @@
+    [DataField]
+    public string[] SpriteRsiPaths = [];
+
+    /// <summary>
+    /// A dictionary of entity sprite layers that have their
+    /// RSI paths updated when the pipe layer changes.
+    /// </summary>
```
Based on this comment alone, I don't understand how this is different from `SpriteRsiPaths`

chromiumboy replied:
If SpriteRsiPaths is set, the RSI path for the whole sprite is changed. If SpriteLayersRsiPaths is set, only the specified layer has it RSI path changed

Partmedia commented on Content.Shared/Atmos/Components/AtmosPipeLayersComponent.cs:
```diff
@@ -0,0 +1,110 @@
+/// Raised when a player attempts to cycle a pipe to its next layer
+/// </summary>
+[Serializable, NetSerializable]
+public sealed partial class TryCyclingPipeLayerCompletedEvent : SimpleDoAfterEvent;
```
Instead of `TryCycling`, perhaps `TrySetNext` or `TryNext`?

Partmedia commented on Content.Shared/Atmos/Components/PipeAppearanceComponent.cs:
```diff
@@ -5,6 +5,8 @@ namespace Content.Shared.Atmos.Components;
 {
-    [DataField("sprite")]
-    public SpriteSpecifier.Rsi Sprite = new(new("Structures/Piping/Atmospherics/pipe.rsi"), "pipeConnector");
+    [DataField]
+    public SpriteSpecifier.Rsi[] Sprite = [new(new("Structures/Piping/Atmospherics/pipe.rsi"), "pipeConnector"),
+        new(new("Structures/Piping/Atmospherics/pipe_alt1.rsi"), "pipeConnector"),
+        new(new("Structures/Piping/Atmospherics/pipe_alt2.rsi"), "pipeConnector")];
```
Are these sprite paths validated?

Partmedia commented on Content.Shared/Atmos/Consoles/Components/AtmosMonitoringConsoleDeviceComponent.cs:
```diff
@@ -16,6 +16,9 @@ public sealed partial class AtmosMonitoringConsoleDeviceComponent : Component
+    [DataField]
     public ProtoId<NavMapBlipPrototype>? NavMapBlip = null;
+
+    [DataField]
+    public bool ShowAbsentConnections = true;
```
Missing documentation.

Partmedia commented on Content.Shared/Atmos/Consoles/SharedAtmosMonitoringConsoleSystem.cs:
```diff
@@ -15,7 +15,7 @@ public override void Initialize()
     private void OnGetState(EntityUid uid, AtmosMonitoringConsoleComponent component, ref ComponentGetState args)
     {
-        Dictionary<Vector2i, Dictionary<(int, string), ulong>> chunks;
+        Dictionary<Vector2i, Dictionary<(int, byte, string), ulong>> chunks;
```
This tuple `(int, byte, string)` is used in so many different places. It deserves a named type. See DRY principle

Partmedia commented on Content.Shared/Atmos/EntitySystems/SharedAtmosPipeLayersSystem.cs:
```diff
@@ -0,0 +1,246 @@
+
+        if (isCovered || isMissingTool)
+        {
+            var label = "???";
```
Dead store, since one of `isCovered` or `isMissingTool` must be set. Consider reworking logic.

Partmedia commented on Resources/Locale/en-US/atmos/atmos-pipe-layers.ftl:
```diff
@@ -0,0 +1,10 @@
+atmos-pipe-layers-component-current-layer = It is in its {$layerName} configuration.
+atmos-pipe-layers-component-select-layer = {CAPITALIZE($layerName)} configuration
+atmos-pipe-layers-component-tool-missing = Requires {INDEFINITE($toolName)} {$toolName}
+atmos-pipe-layers-component-pipes-are-covered = The pipes are covered
+atmos-pipe-layers-component-cannot-adjust-pipes = You need to uncover the pipes before they can be adjusted.
```
Why not merge these two into the same string?

Partmedia commented on Resources/Locale/en-US/atmos/atmos-pipe-layers.ftl:
```diff
@@ -0,0 +1,10 @@
+atmos-pipe-layers-component-change-layer = Adjusted to its {$layerName} configuration.
+atmos-pipe-layers-component-current-layer = It is in its {$layerName} configuration.
+atmos-pipe-layers-component-select-layer = {CAPITALIZE($layerName)} configuration
+atmos-pipe-layers-component-tool-missing = Requires {INDEFINITE($toolName)} {$toolName}
```
Complete sentence

chromiumboy replied:
Not sure what you mean

chromiumboy commented on Content.Client/Atmos/AlignAtmosPipeLayers.cs:
```diff
@@ -0,0 +1,200 @@
+            var direction = (_eyeManager.CurrentEye.Rotation + gridRotation + Math.PI / 2).GetCardinalDir();
+            var multi = (direction == Direction.North || direction == Direction.South) ? -1f : 1f;
+
+            args.WorldHandle.DrawCircle(worldPosition, GuideRadius, _guideColor);
+            args.WorldHandle.DrawCircle(worldPosition + gridRotation.RotateVec(new Vector2(multi * GuideOffset, GuideOffset)), GuideRadius, _guideColor);
+            args.WorldHandle.DrawCircle(worldPosition - gridRotation.RotateVec(new Vector2(multi * GuideOffset, GuideOffset)), GuideRadius, _guideColor);
```
I wanted something that wasn't too obtrusive. And also something that I didn't have to worry about rotation with - it's a real pain

chromiumboy commented on Content.Server/Atmos/Consoles/AtmosMonitoringConsoleSystem.cs:
```diff
@@ -282,16 +282,11 @@ private bool TryGetAtmosDeviceNavMapData
         var direction = xform.LocalRotation.GetCardinalDir();
+        var netId = TryGettingFirstPipeNode(uid, out var _, out var firstNetId) ? firstNetId : -1;
+        var color = TryComp<AtmosPipeColorComponent>(uid, out var atmosPipeColor) ? atmosPipeColor.Color : Color.White;
+        var layer = TryComp<AtmosPipeLayersComponent>(uid, out var atmosPipeLayers) ? atmosPipeLayers.CurrentPipeLayer : (byte)0;
```
It's the default layer

chromiumboy commented on Content.Server/Atmos/Consoles/AtmosMonitoringConsoleSystem.cs:
```diff
@@ -282,16 +282,11 @@ private bool TryGetAtmosDeviceNavMapData
             return false;
 
         var direction = xform.LocalRotation.GetCardinalDir();
+        var netId = TryGettingFirstPipeNode(uid, out var _, out var firstNetId) ? firstNetId : -1;
```
I haven't been able to find an invalid number for net IDs

chromiumboy commented on Content.Client/Atmos/Consoles/AtmosMonitoringConsoleSystem.cs:
```diff
@@ -15,7 +15,7 @@ public override void Initialize()
 
     private void OnHandleState(EntityUid uid, AtmosMonitoringConsoleComponent component, ref ComponentHandleState args)
     {
-        Dictionary<Vector2i, Dictionary<(int, string), ulong>> modifiedChunks;
```
It's a bitmask look up for atmos pipes, 1 for occupied and 0 for empty. Indexed by the net ID, layer, and color hexcode of the pipe. Based on your other comments I'll turn this into a data struct of some sort

Partmedia commented on Content.Client/Atmos/UI/GasAnalyzerWindow.xaml.cs:
```diff
@@ -136,6 +136,7 @@ public void Populate(GasAnalyzerUserMessage msg)
                 else
                 {
                     // oh shit of fuck its more than 4 this ui isn't gonna look pretty anymore
+                    CDeviceMixes.RemoveAllChildren();
```
Having this explained as a comment would be nice, since otherwise it's not very obvious and someone could accidentally remove it.

Partmedia reviewed (approved):
I'm going to stamp "approved" to satisfy the merge requirements, in case I get too busy again soon. Overall things look reasonable. If you are able to get around to it, I think the top items before merging this (as opposed to merging then iterating later) are:

1. Adding a data struct for the `(int, byte, string)` tuple.
2. Adding a method that does the `TryComp<AtmosPipeLayersComponent>(uid, out var atmosPipeLayers) ? atmosPipeLayers.NumberOfPipeLayers : 1` check.
3. Consider if `byte` or some other type is better suited to hold the layer number.

chromiumboy commented:
> I'm going to stamp "approved" to satisfy the merge requirements, in case I get too busy again soon. Overall things look reasonable. If you are able to get around to it, I think the top items before merging this (as opposed to merging then iterating later) are:
> 
> 1. Adding a data struct for the `(int, byte, string)` tuple.
> 2. Adding a method that does the `TryComp<AtmosPipeLayersComponent>(uid, out var atmosPipeLayers) ? atmosPipeLayers.NumberOfPipeLayers : 1` check.
> 3. Consider if `byte` or some other type is better suited to hold the layer number.

I've addressed most of your comments, the outstanding ones I've left unresolved. If you could let me know what's needed, I can address them as well

Partmedia commented on Content.Client/Atmos/EntitySystems/AtmosPipeAppearanceSystem.cs:
```diff
@@ -61,33 +73,43 @@ private void OnAppearanceChanged(EntityUid uid, PipeAppearanceComponent componen
-            if (!_sprite.LayerMapTryGet((uid, args.Sprite), layerKey, out var key, false))
-                continue;
+            // Extract the pipe direction for the current layer
+            var pipeLayerConnectedDirections = (PipeDirection)(15 & (worldConnectedDirections >> (PipeDirectionHelpers.PipeDirections * i)));
```
I would still prefer a comment or explanation instead of using a magic number. This is like `UINT16_MAX` or something.

Partmedia replied:
But really, this represents something like `ALL_DIRECTIONS_BITMAP`.

chromiumboy replied:
Yes, it's a bit mask that's being used to extract the cardinal pipe orientations of the current layer from worldConnectedDirections. I've updated the comment to add more info

Partmedia commented on Content.Client/Atmos/EntitySystems/AtmosPipeAppearanceSystem.cs:
```diff
@@ -61,36 +72,51 @@ private void OnAppearanceChanged(EntityUid uid, PipeAppearanceComponent componen
 
+    private int GetNumberOfPipeLayers(EntityUid uid, out AtmosPipeLayersComponent? atmosPipeLayers)
+    {
+        return TryComp(uid, out atmosPipeLayers) ? atmosPipeLayers.NumberOfPipeLayers : 1;
```
Is `Resolve` more appropriate here?

chromiumboy replied:
I don't think its needed. Resolve lets you optionally pass in a component into the function, and if you don't supply it, Resolve will get it for you. But in both cases, I don't have the component to pass in optionally

Partmedia replied:
This was a mistake on my part. Somehow I didn't see the `out` part of this function declaration, so I misunderstood what this function did. My bad.

Partmedia commented:
Thanks for the changes. Please go ahead and merge this if you're finished with the last touch-ups.

ArtisticRoomba reviewed (approved):
Oh, it's so beautiful.

VerinSenpai commented:
*hyperventilating intensifies*

AndrewTateOfficialSS14 commented:
Found a bug with this layering code, it says it is unable to place a pipe because it can;t fit over other pipe, but in theory it would fit...
![image](https://github.com/user-attachments/assets/43c202b8-d5c7-46df-a332-dd29054c5cab)

MidZik commented:
> Found a bug with this layering code, it says it is unable to place a pipe because it can;t fit over other pipe, but in theory it would fit...

It's ambiguous which device should receive input. Both the manifold and the pipe bend want input from the secondary config in the tile above.

Unrelated, are there any plans to allow multiple pipe ghosts to be placed on the same tile if they're in different configurations?
